File: Anna/ObjC/ANAAnalyzable.m
//
//  ANAAnalyzable.m
//  Anna
//
//  Created by William on 01/07/2017.
//
//

#import "ANAAnalyzable.h"
#import <Anna/Anna-Swift.h>

@implementation NSObject (ANAPrefix)
        
- (id<ANAPrefix> (^)(SEL))ana_context {
    id __unsafe_unretained target = self;
    return ^id(SEL selector) {
        return [[ANAPrefix alloc] initWithTarget:target
                                        selector:selector];
    };
}

- (id<ANAManager>)analyticsManager {
    return ANAManager.sharedManager;
}

@end

File: Anna/ObjC/ANAAnalyzable.swift
//
//  ANAAnalyzable.swift
//  Anna
//
//  Created by William on 12/07/2017.
//
//

import Foundation

extension
NSObject {
    public func
        ana_context() ->(Selector)->ANAPrefixing {
        return { [unowned self] (selector) in
            return ANAPrefix(
                target: self as! ANAAnalyzable,
                selector: selector
            )
        }
    }

    public func
        ana_analyticsManager() ->ANAEventDispatching {
        return ANAManager.sharedManager
    }
}
